doc/man1/bump.1
.\"
.\" $Id: bump.1,v 1.1 2000/09/14 21:54:44 nstraz Exp $
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\" 
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\" 
.\" This program is distributed in the hope that it would be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\" 
.\" Further, this software is distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\" 
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc., 59
.\" Temple Place - Suite 330, Boston MA 02111-1307, USA.
.\" 
.\" Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
.\" Mountain View, CA  94043, or:
.\" 
.\" http://www.sgi.com 
.\" 
.\" For further information regarding this notice, see: 
.\" 
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan/
.\"
.TH BUMP 1 "14 Sep 2000" "LTP" "Linux Test Project"
.SH NAME
bump \- send signal to tags run by pan
.SH SYNOPSIS
\fBbump [-1] [-s \fIsig\fB] [\fI-a active-file\fB] [tags...]
.SH DESCRIPTION

Bump will send a SIGINT signal to processes, given that each process has a
corresponding tag in an active-file.  The active-file is the same one that is
used by the pan to start the processes.

If the active file has multiple occurrences of a single tag name then only the
first process will be signaled.  You may specify the tag name multiple times
on the commandline if necessary.

.TP 1i
\fB-1\fP
Send a SIGUSR1 signal.  By default a SIGINT will be sent.
.TP 1i
\fB-a \fIactive_file\fB
A file containing the tagnames, pids, and commands being run by a pan.  If this
is not specified then the ZOO environment variable will be read for the name of
the directory where the active file can be found.
.TP 1i
\fB-s \fIsig\fB
Used to specify a signal number to send.  By default a SIGINT will be sent.

.in -1i

.SH ENVIRONMENT
.TP
ZOO
If set, should name the directory where the active file can be found.
This is ignored if \fI-a\fP is specified.

.SH FILES
.TP
active
Default name of active file if \fI-a\fP is not specified.  This is prefixed
by the directory name found in the ZOO environment variable.

.SH "SEE ALSO"
Zoo tools - pan(1)

.SH DIAGNOSTICS
Exits zero, unless it cannot find the active file or if there were no tags
listed on the commandline.


ChangeLog
LTP-20020307
------------
o	fixed IDcheck.sh to prompt for IDs only when missing (Jay Huie)
o	added time duration option to pan (Jay Huie)
o	fixed bug [ #516577 ] ftruncate03 test (Jay Huie)
o	added 4 more cases to getcwd02 test (Jay Huie)
o	added time-schedule tool by Richard Gooch under sched test
	section (Manoj Iyer)
o	added trace_sched tool under sched test section (Manoj Iyer)
o	fixed bug [ #511427 ] pread02 test (Paul Larson)
o	fixed bug [ #523055 ] sched_getscheduler test (Paul Larson)
o	fixed bug [ #523137 ] sched_setscheduler02 test (Paul Larson)
o	fixed bug [ #525688 ] sendfile01 (Paul Larson)
o	added sigpending02 test to syscalls section (Paul Larson)
o       created a KNOWN-FAILURES file to document known 
        test failures (Robbie Williamson)
o	added sem02 test to ipc section (Robbie Williamson)
o	added SCTP section to network tests (Robbie Williamson)
o	fixed bug [ #505515 ] perf_lan6 test (Robbie Williamson)
o	added disktest by Brent Yardley to io section (Robbie Williamson)
o	closed the following bugs:
	491285, 491286, 506536, 514408 (Paul Larson and Robbie Williamson)


LTP-20020207
------------
o       added support for cross-compiling (Todd Inglett)
o       added LKML's cache_leak testcase to ltp/scratch (Nate Straz)
o       added IPv6 support (Robbie Williamson)
o       added "gethost" to /tools (Robbie Williamson)
o	fixed the race conditions in the float tests and removed 
	the sleeps (Robbie Williamson)
o       enabled non-root make authority (Paul Larson)
o       separated compilation into "make" and "make install" (Paul Larson)
o       added ipc_stress test (Manoj Iyer)
o       added pthreads_stress test (Manoj Iyer)
o       made changes to support architecture independence (Manoj Iyer & Paul Larson)
o	closed the following bugs: 
	504960, 505108, 504613, 504616, 491283, 506689, 508055, 506692, 508074
	491289, 506662, 511383, 511391, 511427, 511494, 504649, 514050, 512953
	(Manoj Iyer, Paul Larson, and Robbie Williamson)	


LTP-20020108
------------
o	fixed IDcheck.sh to work with bash1 (Nate Straz)
o	updated menu script (Robbie Williamson)
o	reset errno to 0 at the beginning of TEST macro (Paul Larson)
o	added floating point stress tests (Bull/IBM)
o	added fsx-linux NFS tests (Apple)	
o	define GLIBC_SIGACTION_BUG for sigaction02 (Robbie Williamson)
o	removed bogus testcase from dup201 test (Paul Larson)
o	fixed recvfrom01 case 5 (Wayne Boyer)
o	workaround for smp issue with waitpid12 (Manfred Spraul)


LTP-20011206
------------
o	necessary users/groups can now be created on installation with 
	user's permission
o	added a simple menu-based interface for running the LTP testsuite
o	fixed negative duration in pan output when -l isn't used
o	new set of tests under fs-maim
o	fixed race condition in nfslock01
o	ar01, ld01, ldd01, nm01, objdump01, and size01 fixed when multiple 
	copies are run simultaneously
o	workaround for SIGTTOU hang in ioctl02
o	shmget03 uses IPC_PRIVATE to make it safe when running multiple copies
o	compiler warnings and other minor errors fixed in many tests

LTP-20011107
------------
o	Many improvements to mtest05 and mtest06 tests
o	new test nfsstress
o	included ver_linux in LTP and made it run at the end of test scripts
o	check for necessary users/groups in Makefile and warn if they don't
	exist
o	documented the users and groups necessary for the testcases to run
o	simplified telnet01 when looking for root prompt
o	removed incorrect testcase from sendfile03
o	fixed modify_ldt01 test problems on newer kernels
o	added setrlimit to unlimit core size in waitpid05 test to work 
	around systems where this is set to 0
o	fixed mmstress pthread hang problem

LTP-20010925
------------
o	Testcases have been organized into categories in a directory 
	structure under the testcases subdirectory
o	About 400 new testcases added covering various system calls
o	Andreas Jaeger got confirmation from the glibc people that the
	ulimit test was broken as a result of some changes made in 
	glibc2.2.5.  Since it's the test that's broken, Nate Straz 
	removed the offending testcase.
o	Updated README and INSTALL files
o	Descriptions of the testcases have been added to the doc 
	directory and are organized by type (syscalls.txt, fs.txt, etc.)

------------
2000-09-18 Juan J. Quintela <quintela@fi.udc.es>
           Aaron Laffin <alaffin@sgi.com>

  * tests/mmap001.c: first checkin of a modified version of
  mmap001.c from the memtest suite.

2000-09-08  Egor Duda <deo@logos-m.ru>

	* tests/fpathconf01.c: Close opened file at exit.
	* tests/select01.c: Ditto.

2000-09-08  Andrea Arcangeli <andrea@suse.de>
  * tests/fcntl09.c, tests/fcntl10.c: Rearrange the F_SETLK test to
  to correctly use F_RDLCK and F_WRLCK.

2000-09-06  Egor Duda  <deo@logos-m.ru>

	* lib/parse_opts.c (parse_opts): Initialize allocated string
	to prevent heap corruption.



pan/bump.c
/*
 * Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it would be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * Further, this software is distributed without any warranty that it is
 * free of the rightful claim of any third person regarding infringement
 * or the like.  Any license provided herein, whether implied or
 * otherwise, applies only to this software file.  Patent licenses, if
 * any, provided herein do not apply to combinations of this program with
 * other software, or any other product whatsoever.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write the Free Software Foundation, Inc., 59
 * Temple Place - Suite 330, Boston MA 02111-1307, USA.
 *
 * Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
 * Mountain View, CA  94043, or:
 *
 * http://www.sgi.com
 *
 * For further information regarding this notice, see:
 *
 * http://oss.sgi.com/projects/GenInfo/NoticeExplan/
 *
 */
/* $Id: bump.c,v 1.2 2000/09/21 20:42:31 nstraz Exp $ */
#include <stdio.h>
#include <errno.h>
#include <sys/signal.h>
#include <string.h>

#include "zoolib.h"

pid_t
read_active( FILE *fp, char *name );

static char *errmsg;

main( int argc, char **argv ){
	extern char *optarg;
	extern int optind;
	int c;
	int x;
	char *active = NULL;
	pid_t nanny;
	FILE *fp;
	int sig = SIGINT;

	while( (c = getopt(argc, argv, "a:s:12")) != -1 ){
		switch(c){
			case 'a':
				active = (char*)malloc(strlen(optarg)+1);
				strcpy( active, optarg );
				break;
			case 's':
				sig = atoi( optarg );
				break;
			case '1':
				sig = SIGUSR1;
				break;
			case '2':
				sig = SIGUSR2;
				break;
		}
	}

	if( active == NULL ){
		active = zoo_active();
		if( active == NULL ){
			fprintf(stderr, "bump: Must supply -a or set ZOO env variable\n");
			exit(1);
		}
	}
	if( optind == argc ){
		fprintf( stderr, "bump: Must supply names\n");
		exit(1);
	}

	/* need r+ here because we're using write-locks */
	if( (fp = open_file( active, "r+", &errmsg )) == NULL ){
		fprintf(stderr, "bump: %s\n", errmsg);
		exit(1);
	}
	while( optind < argc ){
		/*printf("argv[%d] = (%s)\n", optind, argv[optind] );*/
		nanny = read_active( fp, argv[optind] );
		if( nanny == -1 ){
			fprintf(stderr, "bump: Did not find name '%s'\n",
				argv[optind] );
		}
		else{
			if( kill( nanny, sig ) == -1 ){
				if( errno == ESRCH ){
					fprintf(stderr,"bump: Did not find nanny '%s', pid %d\n",
						argv[optind], nanny );
					if( clear_active( fp, nanny, &errmsg ) != 1 )
						fprintf(stderr,"bump: %s\n", errmsg);
				}
			}
		}
		++optind;
	}
	fclose( fp );

	exit(0);
}




pid_t
read_active( FILE *fp, char *name )
{
	char buf[1024];
	pid_t nanny = -1;
	char *n;
	int len;

	if( lock_file( fp, F_WRLCK, &errmsg ) == -1 ){
		fprintf(stderr, "bump: %s\n", errmsg );
		return(-1);
	}
	if( seek_file( fp, 0, SEEK_SET, &errmsg ) == -1 ){
		fprintf(stderr, "bump: %s\n", errmsg );
		return(-1);
	}

	len = strlen( name );
	while(1){
		if( fgets( buf, 1023, fp ) == NULL ){
			/*printf("end of file\n");*/
			break;
		}

		if( buf[0] == '#' ){
			/*printf("skip line (%s)\n", buf );*/
			continue;
		}

		/* get name */
		if( (n = strchr( buf, ',' )) == NULL ){
			/*printf("no comma (%s)\n", buf );*/
			continue;
		}

		if( strncmp( n + 1, name, len ) != 0 ){
			/*printf("not matching (%s) and (%s)\n",
			       n+1, name );*/
			continue;
		}

		nanny = atoi( buf );
		break;
	}/* while */

	if( lock_file( fp, F_UNLCK, &errmsg ) == -1 ){
		fprintf(stderr, "bump: %s\n", errmsg );
		return(-1);
	}
	return nanny;
}

